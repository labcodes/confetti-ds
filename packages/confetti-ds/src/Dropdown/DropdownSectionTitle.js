import React, { Component } from "react";
import { PropTypes } from "prop-types";

export default class DropdownSectionTitle extends Component {
  static propTypes = {
    /** This is DropdownSectionTitle's title */
    text: PropTypes.node.isRequired,
    /** This is DropdownSectionTitle's color. The color is inherit from dropdown color  */
    color: PropTypes.string,
  };

  static defaultProps = {
    color: "teal",
  };

  render() {
    const { text, color } = this.props;
    return (
      <span
        className={`lab-dropdown__section-title lab-dropdown__section-title--color-${color}`}
      >
        {text}
      </span>
    );
  }
}
