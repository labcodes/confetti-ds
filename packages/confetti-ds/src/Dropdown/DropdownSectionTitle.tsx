import React from "react";

export interface DropdownSectionTitleProps {
  /** This is DropdownSectionTitle's title */
  text: React.ReactNode;
  /** This is DropdownSectionTitle's color. The color is inherit from dropdown color  */
  color?: "white" | "mineral" | "teal" | "purple";
}
export default function DropdownSectionTitle({
  text,
  color = "teal",
}: DropdownSectionTitleProps) {
  return (
    <span
      className={`lab-dropdown__section-title lab-dropdown__section-title--color-${color}`}
    >
      {text}
    </span>
  );
}
