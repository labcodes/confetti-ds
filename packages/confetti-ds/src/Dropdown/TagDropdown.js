import React, { Component } from "react";
import { PropTypes } from "prop-types";

import AbstractDropdown from "./AbstractDropdown";

export default class TagDropdown extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    text: PropTypes.string,
    color: PropTypes.string,
  };

  static defaultProps = {
    text: "foo",
    color: "teal",
  };

  render() {
    const { children, text, color } = this.props;
    return (
      <AbstractDropdown dropdownType="tag" text={text} color={color}>
        {children}
      </AbstractDropdown>
    );
  }
}
