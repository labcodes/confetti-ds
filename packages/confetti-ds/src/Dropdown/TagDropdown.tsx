import React, { SyntheticEvent } from "react";

import AbstractDropdown from "./AbstractDropdown";

export interface TagDropdownProps {
  /** Function called after the dropdown is opened. */
  onOpen?: (event: SyntheticEvent) => any;
  /** Function called after the dropdown is closed. */
  onClose?: (event: SyntheticEvent) => any;
  /** Function called when the user selects one valid option. */
  onSelect: (event: SyntheticEvent) => any;
  /** This is the dropdown of children. It can be a TagDropdownItem or a SectionTitle, and if the user doesn't pass a child, it will throw an error. */
  children: JSX.Element | JSX.Element[];
  /** This is the dropdown color, and it will set the color of its section title and trigger. */
  color?: "white" | "mineral" | "teal" | "purple";
  /** This is the dropdown default trigger title. It will mount with this default text until the user selects an option. */
  defaultText: string;
  /** This is the dropdown id. It requires a unique id. */
  id: string;
  /** This is the controlled value of the TagDropdown */
  value?: string | number;
}

export default function TagDropdown({
  onOpen = () => {},
  onClose = () => {},
  onSelect,
  children,
  color = "teal",
  defaultText,
  id,
  value = "",
}: TagDropdownProps) {
  return (
    <AbstractDropdown
      dropdownType="tag"
      defaultText={defaultText}
      color={color}
      onOpen={onOpen}
      onClose={onClose}
      onSelect={onSelect}
      id={id}
      value={value}
    >
      {children}
    </AbstractDropdown>
  );
}
