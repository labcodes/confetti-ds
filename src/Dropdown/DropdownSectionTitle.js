import React, { Component } from "react";
import { PropTypes } from "prop-types";

export default class DropdownSectionTitle extends Component {
  static propTypes = {
    text: PropTypes.node.isRequired,
    color: PropTypes.string,
  };

  static defaultProps = {
    color: "teal",
  };

  render() {
    const { text, color } = this.props;
    return (
      <span
        className={`lab-dropdown__section-title lab-dropdown__section-title--color-${color}`}
      >
        {text}
      </span>
    );
  }
}
