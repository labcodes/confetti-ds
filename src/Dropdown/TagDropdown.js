import React, { Component } from "react";
import { PropTypes } from "prop-types";

import AbstractDropdown from "./AbstractDropdown";

export default class TagDropdown extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    text: PropTypes.string.isRequired,
    color: PropTypes.string,
    onOpen: PropTypes.func,
    onClose: PropTypes.func,
  };

  static defaultProps = {
    color: "teal",
    onOpen: () => {},
    onClose: () => {},
  };

  render() {
    const { children, text, color, onOpen, onClose } = this.props;
    return (
      <AbstractDropdown
        dropdownType="tag"
        text={text}
        color={color}
        onOpen={onOpen}
        onClose={onClose}
      >
        {children}
      </AbstractDropdown>
    );
  }
}
