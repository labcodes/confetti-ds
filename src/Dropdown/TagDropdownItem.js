import React, { useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import { isEmpty } from "lodash";

import Icon from "../Icon";
import { ICON_TYPES, TAG_COLORS } from "../constants";
import AbstractTag from "../Tags/AbstractTag";

export default function TagDropdownItem({
  text,
  thumbSrc,
  icon,
  color,
  skin,
  isOutline,
  onClick,
  disabled,
  tabIndex,
  onInteraction,
  setRef,
  value,
  isSelected,
}) {
  const checkThumbAndIcon = (thumbSrc, icon) => {
    const errorMessage =
      "`TagDropdownItem` can't be initialized with both `thumb` and `icon` props.";
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw new Error(errorMessage);
    }
  };

  // super constructor
  const handleCheckThumbAndIcon = useCallback(() => {
    checkThumbAndIcon();
  }, [icon, thumbSrc]);

  // componentDidUpdate
  useEffect(() => {
    handleCheckThumbAndIcon();
  }, []);

  const renderThumb = ({ thumbSrc }) =>
    thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt="" />
    ) : undefined;

  const renderIcon = ({ icon }) =>
    icon ? (
      <Icon
        type={icon}
        color="black-75"
        size="petit"
        className="lab-tag--left-icon"
      />
    ) : undefined;

  return (
    <AbstractTag
      text={text}
      className={`lab-tag--togglable${icon ? ` lab-tag--has-left-icon` : ""}${
        thumbSrc ? ` lab-tag--has-thumb` : ""
      }`}
      isOutline={isOutline}
      skin={skin}
      color={color}
      renderPrefix={renderIcon(icon) || renderThumb(thumbSrc)}
      onClick={onClick}
      disabled={disabled}
      tabIndex={tabIndex}
      onInteraction={onInteraction}
      isDropdown
      setRef={setRef}
      value={value}
      isSelected={isSelected}
      role="option"
    />
  );
}

TagDropdownItem.propTypes = {
  /** This is the Tag's text. */
  text: PropTypes.string.isRequired,
  /** Disables the Tag. Won't be read by screen readers. */
  disabled: PropTypes.bool,
  /** Action to be executed when the Tag is clicked. */
  onClick: PropTypes.func,
  /** Source of the thumb to be rendered. Won't render a thumb if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  thumbSrc: PropTypes.string,
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. Can't have both 'icon' and 'thumbSrc' at the same time. */
  icon: PropTypes.oneOf(ICON_TYPES),
  /** Sets Tag's color. */
  color: PropTypes.oneOf(TAG_COLORS),
  /** Skin of the rendered Tag. */
  skin: PropTypes.oneOf(["pale", "vivid"]),
  /** Sets an outline style. */
  isOutline: PropTypes.bool,
  /** Option tab index */
  tabIndex: PropTypes.string,
  /** This function is used on AbstractTag's componenentDidMount to set the current Ref */
  setRef: PropTypes.func,
  /** This function is used to handle click or keydown interactions */
  onInteraction: PropTypes.func,
  /** This is the option's value */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  /** This prop is a boolean to verify if the option is current selected  */
  isSelected: PropTypes.bool,
};

TagDropdownItem.defaultProps = {
  thumbSrc: "",
  icon: undefined,
  isOutline: false,
  skin: "pale",
  color: undefined,
  disabled: false,
  onClick: () => {},
  setRef: () => {},
  onInteraction: () => {},
  tabIndex: "0",
  isSelected: false,
};
