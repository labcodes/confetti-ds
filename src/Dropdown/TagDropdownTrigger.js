import React from "react";
import AbstractTag from "../Tags/AbstractTag";
import { DropdownTag } from "../Tags";

export default class TagDropdownTrigger extends DropdownTag {
  render() {
    const {
      text,
      icon,
      color,
      skin,
      isOutline,
      disabled,
      ariaDisabled,
      onClick,
      tabIndex,
    } = this.props;
    return (
      <AbstractTag
        className={`lab-tag--dropdown${`${
          icon ? ` lab-tag--has-left-icon` : ""
        }`}`}
        text={text}
        icon={icon}
        color={color}
        skin={skin}
        isOutline={isOutline}
        disabled={!ariaDisabled && disabled}
        ariaDisabled={ariaDisabled}
        onClick={onClick}
        renderPrefix={this.icon()}
        renderSuffix={this.dropdownIcon()}
        tabIndex={tabIndex}
      />
    );
  }
}
