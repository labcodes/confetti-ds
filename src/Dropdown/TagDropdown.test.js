import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import DropdownSectionTitle from "./DropdownSectionTitle";
import TagDropdown from "./TagDropdown";
import TagDropdownItem from "./TagDropdownItem";
import TagDropdownTrigger from "./TagDropdownTrigger";

describe("TagDropdown", () => {
  it("not expected children", () => {
    const onSelectMock = jest.fn();

    const wrapper = mount(
      <TagDropdown
        id="testId"
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
      >
        <div>Just some child 1</div>
        <div>Just some child 2</div>
        <div>Just some child 3</div>
        <div>Just some child 4</div>
      </TagDropdown>
    );

    const childrenLength = wrapper
      .find(".lab-dropdown__content")
      .children().length;

    expect(childrenLength).toBe(0);
  });

  it("expected children", () => {
    const onSelectMock = jest.fn();

    const wrapper = mount(
      <TagDropdown
        color="mineral"
        defaultText="Click me"
        onSelect={onSelectMock}
        id="idTest"
      >
        <div>Just some child 1</div>
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
        <div>Just some child 2</div>
      </TagDropdown>
    );

    const childrenLength = wrapper
      .find(".lab-dropdown__content")
      .children().length;

    expect(childrenLength).toBe(4);
  });

  it("renders with base props", async () => {
    const onSelectMock = jest.fn();

    const renderedComponent = renderer
      .create(
        <TagDropdown id="idTest" defaultText="Click me" onSelect={onSelectMock}>
          <DropdownSectionTitle text="First Section" />
          <TagDropdownItem value="1" text="One" />
          <TagDropdownItem value="2" text="Two" />
          <TagDropdownItem value="3" text="Three" />
        </TagDropdown>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing color as teal", async () => {
    const onSelectMock = jest.fn();

    const renderedComponent = renderer
      .create(
        <TagDropdown
          id="idTest"
          defaultText="Click me"
          color="teal"
          onSelect={onSelectMock}
        >
          <DropdownSectionTitle text="First Section" />
          <TagDropdownItem value="1" text="One" />
          <TagDropdownItem value="2" text="Two" />
          <TagDropdownItem value="3" text="Three" />
        </TagDropdown>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <TagDropdown
        id="idTest"
        defaultText="Click me"
        color="teal"
        onSelect={onSelectMock}
      >
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
      </TagDropdown>
    );
    expect(mountedComponent.find(TagDropdownTrigger).prop("color")).toEqual(
      "teal"
    );
    expect(mountedComponent.find(DropdownSectionTitle).prop("color")).toEqual(
      "teal"
    );
  });

  it("calls props.onClose and props.onOpen when clicked or blured", async () => {
    const mockOnClose = jest.fn();
    const mockOnOpen = jest.fn();
    const onSelectMock = jest.fn();

    const shallowDropdown = mount(
      <TagDropdown
        id="testId"
        dropdownType="tag"
        defaultText="Click me"
        onClose={mockOnClose}
        onOpen={mockOnOpen}
        onSelect={onSelectMock}
      >
        <DropdownSectionTitle text="First Section" />
        <TagDropdownItem value="1" text="One" />
        <TagDropdownItem value="2" text="Two" />
        <TagDropdownItem value="3" text="Three" />
      </TagDropdown>
    );

    const trigger = shallowDropdown.find(TagDropdownTrigger).at(0);
    trigger.simulate("click");

    expect(mockOnOpen.mock.calls.length).toBe(1);
    expect(mockOnClose.mock.calls.length).toBe(0);

    trigger.simulate("click");

    expect(mockOnOpen.mock.calls.length).toBe(1);
    expect(mockOnClose.mock.calls.length).toBe(1);
  });
});
