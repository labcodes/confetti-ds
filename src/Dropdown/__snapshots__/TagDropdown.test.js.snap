// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagDropdown renders as expected when passing color as teal 1`] = `
<div
  className="lab-dropdown"
  onBlur={[Function]}
>
  <div>
    <button
      aria-controls="menu--idTest"
      aria-haspopup={true}
      className="lab-dropdown__invisible-button lab-dropdown__invisible-button--trigger"
      id="menu-button--menu--idTest"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me
    </button>
    <span
      className="lab-tag lab-tag--dropdown lab-tag--teal-pale"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="-1"
    >
      Click me
      <span
        className="lab-tag__dropdown-icon-wrapper"
      >
        <span
          className="lab-icon lab-icon--dropdown-closed lab-icon--black-75 lab-icon--petit lab-tag--dropdown-icon"
        />
      </span>
    </span>
  </div>
  <div
    aria-labelledby="menu-button--menu--idTest"
    className="lab-dropdown__content lab-dropdown__listbox "
    id="menu--idTest"
    role="menu"
    tabIndex="-1"
  >
    <span
      className="lab-dropdown__section-title lab-dropdown__section-title--color-teal"
    >
      First Section
    </span>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--1"
    >
      <button
        aria-disabled={false}
        className="lab-dropdown__invisible-button lab-dropdown__invisible-button--option"
        disabled={false}
        id="option-1--menu--idTest"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        One
      </button>
      <span
        className="lab-tag  lab-tag--pale"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="-1"
      >
        One
      </span>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--2"
    >
      <button
        aria-disabled={false}
        className="lab-dropdown__invisible-button lab-dropdown__invisible-button--option"
        disabled={false}
        id="option-2--menu--idTest"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Two
      </button>
      <span
        className="lab-tag  lab-tag--pale"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="-1"
      >
        Two
      </span>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--3"
    >
      <button
        aria-disabled={false}
        className="lab-dropdown__invisible-button lab-dropdown__invisible-button--option"
        disabled={false}
        id="option-3--menu--idTest"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Three
      </button>
      <span
        className="lab-tag  lab-tag--pale"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="-1"
      >
        Three
      </span>
    </div>
  </div>
</div>
`;

exports[`TagDropdown renders with base props 1`] = `
<div
  className="lab-dropdown"
  onBlur={[Function]}
>
  <div>
    <button
      aria-controls="menu--idTest"
      aria-haspopup={true}
      className="lab-dropdown__invisible-button lab-dropdown__invisible-button--trigger"
      id="menu-button--menu--idTest"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me
    </button>
    <span
      className="lab-tag lab-tag--dropdown lab-tag--teal-pale"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="-1"
    >
      Click me
      <span
        className="lab-tag__dropdown-icon-wrapper"
      >
        <span
          className="lab-icon lab-icon--dropdown-closed lab-icon--black-75 lab-icon--petit lab-tag--dropdown-icon"
        />
      </span>
    </span>
  </div>
  <div
    aria-labelledby="menu-button--menu--idTest"
    className="lab-dropdown__content lab-dropdown__listbox "
    id="menu--idTest"
    role="menu"
    tabIndex="-1"
  >
    <span
      className="lab-dropdown__section-title lab-dropdown__section-title--color-teal"
    >
      First Section
    </span>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--1"
    >
      <button
        aria-disabled={false}
        className="lab-dropdown__invisible-button lab-dropdown__invisible-button--option"
        disabled={false}
        id="option-1--menu--idTest"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        One
      </button>
      <span
        className="lab-tag  lab-tag--pale"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="-1"
      >
        One
      </span>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--2"
    >
      <button
        aria-disabled={false}
        className="lab-dropdown__invisible-button lab-dropdown__invisible-button--option"
        disabled={false}
        id="option-2--menu--idTest"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Two
      </button>
      <span
        className="lab-tag  lab-tag--pale"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="-1"
      >
        Two
      </span>
    </div>
    <div
      className="lab-dropdown__option "
      id="lab-dropdown__option--3"
    >
      <button
        aria-disabled={false}
        className="lab-dropdown__invisible-button lab-dropdown__invisible-button--option"
        disabled={false}
        id="option-3--menu--idTest"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Three
      </button>
      <span
        className="lab-tag  lab-tag--pale"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="-1"
      >
        Three
      </span>
    </div>
  </div>
</div>
`;
